src/collectors/hadoop2/hadoop.py

```diff
@@ -0,0 +1,114 @@
+
+#### Dependencies
+
+ * hadoop
+
+"""
+
+import glob
+import os
+import re
+
+from diamond.metric import Metric
+from diamond.collector import Collector, str_to_bool
+
+
+NEW_CONFIGS = {
+    'metric_files': {
+        'default': ['/var/log/hadoop/*-metrics.out'],
+        'help': "List of paths to process metrics from."
+    },
+    'truncate': {
+       'default': False,
+        'help': "Truncate the metrics files after reading them."
+    }
+}
+
+
+class HadoopMetrics2Collector(Collector):
+
+    re_log = re.compile(r'^(?P<timestamp>\d+) (?P<name>\S+): (?P<metrics>.*)$')
+
+    def _get_new_configs(self, val):
+        return dict(map(lambda (a,b): (a, b[val]), NEW_CONFIGS.items()))
+
+    def get_default_config_help(self):
+        config_help = super(HadoopMetrics2Collector, self).get_default_config_help()
+        config_help.update(self._get_new_configs('help'))
+        return config_help
+
+    def get_default_config(self):
+        config = super(HadoopMetrics2Collector, self).get_default_config()
+        config.update(self._get_new_configs('default'))
+        return config
+
+    def collect(self):
+        metrics = self.config['metric_files']
+        if not isinstance(metrics, list):
+            metrics = [str(metrics)]
+
+        for pattern in metrics:
+            for filename in glob.glob(pattern):
+                self._collect_from(filename)
+
+    def _collect_from(self, filename):
+        """Loops through each line in the file, parses the logged data into
+        metrics, and publishes the metrics."""
+        if not os.access(filename, os.R_OK):
+            self.log.error('HadoopMetrics2Collector unable to read "%s"', filename)
+            return
+
+        if self.config['truncate']:
+            _file = open(filename, 'r+')
+        else:
+            _file = open(filename, 'r')
+
+        self.config['hostname_method'] = 'uname_short'
+
+        for line in _file:
+            match = self.re_log.match(line)
+            if not match:
+                continue
+            raw_data = match.groupdict()
+
+            metrics = {}
+            extra_data = {}
+            for metric in raw_data['metrics'].split(','):
+                metric = metric.strip()
+                if '=' in metric:
+                    key, value = metric.split('=', 1)
+                    try:
+                        metrics[key] = round(float(value))
+                    except ValueError:
+                        extra_data[key] = value
+
+            host = extra_data.get('Hostname', None) or self.get_hostname()
+            partial_path = 'hadoop.{}.'.format(raw_data['name'])
+
+            for key, val in metrics.items():
+                full_path = partial_path + key
+                self._publish(key, val, full_path, host, raw_data['timestamp'])
+
+        if str_to_bool(self.config['truncate']):
+            _file.seek(0)
+            _file.truncate()
```

Comment:
One concern about this. What happens if data gets written between when we reach current EOF and when we go to truncate the file?

We would accidentally delete lines we havent drained yet? Or does the file handle only truncate what we have read. Can you do a quick test to verify... if we have to worry about this, we can track the size of the lines as we read them, and then truncate from start with the size parameter to truncate(), which should work.

Reply:
I modified the code to first get the file size as soon as the file is open:
```
_file = open(filename, 'r')
    original_file_end = _get_file_end(_file)
```
then when we're finished reporting metrics, truncate the file up to this size:
```
with open(filename, 'r+') as _file:
    new_file_end = _get_file_end(_file)
     _file.truncate(new_file_end - original_file_end)
```